### ISSUE TITLE
Add toEncoding/fromEncoding methods to ToAvro/FromAvro.

### ISSUE THREAD

**Issue by TikhonJelvis** [haskell-works/avro#87]
Right now, we always use `Value` as an intermediate type for encoding and decoding Avro values. 

Using an intermediate type like this for larger amounts of data is _expensive_. On an internal project, I found that using an intermediate representation like `Value` was about 3x slower and used about *40×* more memory than going directly to/from normal Haskell types generated by TH.

Aeson has the same problem, with similar time/space overhead for using `Aeson.Value` on large inputs. Aeson solves this problem by having a `toEncoding` function as part of `ToJSON` that generates a bytestring directly rather than going through `Aeson.Value`. 

We should add similar `toEncoding` and `fromEncoding` methods to `ToAvro` and `FromAvro`, and generate implementations for these methods for our TH types. I recently implemented this for some slightly different TH types on an internal project at Target and the logic was a lot simpler than I had thought—in fact, the logic to go directly to binary turned out to be _simpler_ than going through an intermediate type! This change also removed a massive memory leak and improved serialization performance significantly.

**Comment by AlexeyRaga** [comment]
Thanks for bringing it into the light!
I was thinking about something like this for a while, but haven't had a chance to look at it properly.

One complication for me was the ability to decode an Avro container/message with "smaller" (but compatible) schema: the direct decoders should know how to skip bits that are not used. So some knowledge of deconflicting is now required at the decoding layer.
 
I am sure that it is solvable, but, as I said, I didn't have time to think of a proper solution.
I am very curious about your ideas and findings!

**Comment by AlexeyRaga** [comment]
An update (//cc @TikhonJelvis @dwincort)

I am slowly working in this direction and the results are quite impressive:
```
benchmarked Decoding container/Avro Strict
time                 334.6 ms   (332.8 ms .. 335.8 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 335.3 ms   (334.6 ms .. 336.3 ms)
std dev              1.377 ms   (850.8 μs .. 2.023 ms)

benchmarked Decoding container/Avro Lazy
time                 207.5 ms   (206.3 ms .. 208.9 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 208.0 ms   (207.1 ms .. 208.8 ms)
std dev              1.561 ms   (989.3 μs .. 2.479 ms)

benchmarked Decoding container/From Encoding
time                 87.77 ms   (86.61 ms .. 90.21 ms)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 87.49 ms   (86.40 ms .. 88.21 ms)
std dev              1.544 ms   (1.016 ms .. 2.204 ms)
```

The progress is in this branch: https://github.com/haskell-works/avro/tree/decode-optimisation
and it includes schema-based deconflicting approach that @dwincort initially suggested.

Once it is solid I am thinking of getting rid of both `Strict` and `Lazy` variants of decoding that we currently have since they don't seem to be better than the new "direct" approach.

The biggest issue at the moment is parsing default values from Schema. It is done for the "old" way of doing things, and it should be possible to do it in a "new" way too.

**Comment by AlexeyRaga** [comment]
Addressed in #141
